README.md

```diff
@@ -1,277 +1,899 @@
+Swap "writers" must first create a swap by using the `CreateSwap` Redeemer in the following manner:
 
-When engaging in swaps, it is only necessary that the desired swap ratio is met; **not all assets at the swap address or UTxO must be swapped.** For example, if there is 100 ADA in a swap address requesting 2:1 for DUST, a user may swap 20 ADA, as long as they return 80 ADA and 10 DUST in the same Tx. Since every user explicitly defines their desired swap ratios, oracles are not required. The "global" price naturally emerges where the local bids and asks meet - just like an order-book.
+In order to mint beacons with this redeemer, **all of the following must be true**:
 
-> **Note**
-> 
-> In previous versions, users could have specified a price in units of ADA and the script would convert it to units of Lovelace during execution. This feature was removed to save on execution costs. Now, all prices for ADA must be in units of Lovelace.
+1) The beacons must go to an addres protected by the two-way swap validator script.
+2) The beacons must go to an address using a valid staking credential.
+3) The UTxOs with the beacons must have the proper value:
+    - Exactly three kinds of beacons: pair beacon, asset1 beacon, and asset2 beacon.
+    - The beacons must correspond to the beacons in the datum.
+    - There must be exactly 1 of each beacon.
+    - The UTxO must have asset1 and/or asset2.
+    - No extraneous assets are in the UTxO. ADA is always allowed.
+4) The beacons must be stored with the proper inline `SwapDatum`:
+    - `beaconId` == this policy id.
+    - `pairBeacon` == trading pair beacon asset name for this swap.
+    - `asset1Id` == asset_id of asset1 for that trading pair.
+    - `asset1Name` == asset_name of asset1 for that trading pair.
+    - `asset1Beacon` == asset1 beacon asset name.
+    - `asset2Id` == asset_id asset2 for that trading pair.
+    - `asset2Name` == asset_name asset2 for that trading pair.
+    - `asset2Beacon` == asset2 beacon asset name.
+    - `forwardPrice` denominator > 0
+    - `forwardPrice` > 0
+    - `reversePrice` denominator > 0
+    - `reversePrice` > 0
+    - asset1 < asset2
+
+The validator will assume that the trading pairs are sorted which is why asset1 must be less than
+asset2. Asset1 and asset2 cannot be the same asset.
+
+Once the beacons are minted to the swap address, the spending script does not allow closing the swap
+*unless* the beacons are being burned. This prevents beacons from being sent to unrelated addresses
+and guarantees the integrity of the off-chain queries.
+
+All swaps must be stored with the proper beacons. Due to this requirement, the absolute minimum
```

Comment:
> I believe it will improve the quality of this material as a reference

I'm disagree with this. The spec is not directed towards the laymen; it is directed towards developers (this is partly why I am okay with people skipping it). With that in mind, if I were going through the spec and trying to implement it myself, I would prefer to have everything I need in one location. Jumping around too much can make it difficult to keep track of where I am in the spec.

>  I believe it will improve ... its readability

I agree with this. Duplication can make a document feel dry. To me, there is a trade-off between being a good reference material and being easy to read. I don't think you can have both.

@zhekson1 and I have gone back and forth on a similar topic: since there is significant overlap in the architecture of the P2P protocols and therefore, significant overlap in the features, should those things be extracted out into a separate location like zhekson's defi-protocols [repo](https://github.com/zhekson1/CSL-DeFi-Protocols)? Personally, I am against extracting things out; I would rather duplicate things to have a single location with everything needed to understand it. I think part of the problem of showing people "Why Cardano?" is similar: there is no single location that has what you need. You need to jump around a lot to get the full picture.

Reply:
I decided to keep the duplication in the spec for the reasons outlined above. If it does become to difficult to maintain, I can change it in the future.